rpk: add `rpk cluster partitions list`

r-vasquez opened the pull request:
This PR introduces a new `rpk cluster partitions list` that uses the recently introduced admin API endpoint (https://github.com/redpanda-data/redpanda/pull/14461)

**EXAMPLES**
```
# Listing all
$ rpk cluster partitions list -a                  
NAMESPACE       TOPIC         PARTITION  LEADER-ID  REPLICA-CORE   DISABLED
kafka           bar           0          -          [1-4]          false
kafka           baz           0          -          [0-1 1-1 2-1]  false
kafka           baz           2          -          [0-3 1-3 2-3]  false
kafka           foo           0          -          [0-4]          false
kafka_internal  id_allocator  0          -          [0-5 1-5 2-4]  false


# Filtering by topic
$ rpk cluster partitions list baz   
NAMESPACE  TOPIC  PARTITION  LEADER-ID  REPLICA-CORE   DISABLED
kafka      baz    0          -          [0-1 1-1 2-1]  false
kafka      baz    2          -          [0-3 1-3 2-3]  false

# Format JSON:
$ rpk cluster partitions list baz --format json
[{"ns":"kafka","topic":"baz","partition_id":0,"replicas":[{"node_id":1,"core":6},{"node_id":4,"core":6},{"node_id":5,"core":6}],"disabled":false},{"ns":"kafka","topic":"baz","partition_id":1,"replicas":[{"node_id":0,"core":6},{"node_id":2,"core":6},{"node_id":3,"core":6}],"disabled":false},{"ns":"kafka","topic":"baz","partition_id":2,"replicas":[{"node_id":0,"core":7},{"node_id":4,"core":7},{"node_id":5,"core":7}],"disabled":false},{"ns":"kafka","topic":"baz","partition_id":3,"replicas":[{"node_id":1,"core":7},{"node_id":2,"core":7},{"node_id":3,"core":7}],"disabled":false},{"ns":"kafka","topic":"baz","partition_id":4,"replicas":[{"node_id":0,"core":8},{"node_id":2,"core":8},{"node_id":3,"core":8}],"disabled":false},{"ns":"kafka","topic":"baz","partition_id":5,"replicas":[{"node_id":1,"core":8},{"node_id":4,"core":8},{"node_id":5,"core":8}],"disabled":false}]
```

If the user is trying to hit an old Redpanda version that does not support `/v1/cluster/partition` we query the info to `/v1/partitions` :

```
$ rpk cluster partitions list foo
NAMESPACE  TOPIC  PARTITION  LEADER-ID  REPLICA-CORE  DISABLED
kafka      foo    0          1          [1-5]         -

# And we log:
$ rpk cluster partitions list foo -v
17:06:12.495  DEBUG  Sending request  {"method": "GET", "url": "http://127.0.0.1:9644/v1/cluster/partitions/kafka/foo?disabled=false", "bearer": false, "basic": false}
17:06:12.496  WARN  Admin API 'GET /v1/cluster/partitions' returned 404, trying now with 'GET /v1/partitions'
17:06:12.496  DEBUG  Sending request  {"method": "GET", "url": "http://127.0.0.1:9644/v1/partitions/kafka/foo", "bearer": false, "basic": false}
NAMESPACE  TOPIC  PARTITION  LEADER-ID  REPLICA-CORE  DISABLED
kafka      foo    0          1          [1-5]         -

```
This PR also includes other small unrelated changes that were found during the development such as:
- Adding missing doc and copyright info.
- Changing the log level of a retry error to WARN.
- Renaming a section in a help text for consistency in other rpk commands.

Second part of #14394 

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.2.x
- [ ] v23.1.x
- [ ] v22.3.x

## Release Notes

### Features

* rpk introduces `rpk cluster partitions list` which lets users query cluster-level metadata of all partitions in the cluster.

r-vasquez commented on src/go/rpk/pkg/cli/cluster/partitions/list.go:
```diff
@@ -14,194 +21,107 @@ import (
+				out.MaybeDie(err, "unable to query all partitions in the cluster: %v", err)
+			} else {
+				if withInternals {
+					out.Die("--with-internals not available when listing single topics.\n%v", cmd.UsageString())
```
I'm experimenting here with the error message, printing the error + usage after an incorrect use of flags and/or args is a better alternative to just printing the error.

However, the usageString does not contain the long text, so, wdyt of adding something like `out.ExecAndDie`, we print a message, exec, and die. That way we can:
```
  out.ExecAndDie(cmd.Help, "my message %v", args)
```

But not sure...

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/40799#018bb62a-8572-4979-b63f-f213557e22cd

gene-redpanda commented on src/go/rpk/pkg/cli/cluster/partitions/move_status.go:
```diff
@@ -0,0 +1,216 @@
+				return strconv.Itoa(size)
+			}
+
+			f := func(rr *adminapi.ReconfigurationsResponse) interface{} {
```
I personally think multiple inline functions assigned to variables is tough to read, especially when they aren't being immediately assigned as an input to something else (ex http mux functions to serve endpoints). Any reason to not just have them be standard functions that can be passed in?

r-vasquez replied:
Oh, this is not part of my change (If you see commit by commit, you'll see that I renamed first the `list.go` -> `move_status.go`, then created a new list.go)

But when you see the whole code change, you'll see like I created move_status.go entirely as part of this PR.

gene-redpanda replied:
Ok reasonable enough. I admit to basically never reviewing commit by commit lol

r-vasquez commented on src/go/rpk/pkg/adminapi/api_partition.go:
```diff
@@ -58,6 +60,20 @@ type ReconfigurationsResponse struct {
+	Ns          string              `json:"ns" yaml:"ns"`
+	Topic       string              `json:"topic" yaml:"topic"`
+	PartitionID int                 `json:"partition_id" yaml:"partition_id"`
+	LeaderID    *int                `json:"leader_id,omitempty" yaml:"leader_id,omitempty"`
```
Is the only field listed in the Redpanda Docs that does not come always in the response, if it's not a pointer and instead we use `int`, then it will show 0 when is not there. And 0 is a valid leader_id and it can confuse users.

gene-redpanda replied:
It might be worth leaving a comment to that effect to prevent future changes. I know if I were just looking at the current setup I'd have a nigh uncontrollable urge to make it not a pointer 😂

twmb replied:
+1 to comment, `// LeaderID may be missing in the response.`

r-vasquez replied:
Comment added, thanks.

twmb commented on src/go/rpk/pkg/cli/cluster/partitions/list.go:
```diff
@@ -14,194 +21,107 @@ import (
+				out.MaybeDie(err, "unable to query all partitions in the cluster: %v", err)
+			} else {
+				if withInternals {
+					out.Die("--with-internals not available when listing single topics.\n%v", cmd.UsageString())
```
I'd use the same flag name we use in `rpk topic list`, `-i`, `--internal`

r-vasquez replied:
Ok, but then should we rename `disabled` to `disabled-only` ? `with-internals`  add the internal topics in addition to the rest of the topics. But `disabled` only shows the disabled partitions.

twmb replied:
Actually though I'm not sure --internal is a great flag, I'd maybe rename this to --with-internal (drop the s), rename the `rpk topic list` flag, and hide the old `rpk topic list` flag. I wonder what Confluent tools name this flag.

r-vasquez replied:
In kafka_topics.sh:

```
--exclude-internal                       exclude internal topics when running
                                          list or describe command. The
                                          internal topics will be listed by
                                          default
```

twmb replied:
I'm struggling to think of a scenario where we wouldn't want internal partitions to be displayed -- this is an admin command after all, in service of admins figuring out which partitions are disabled.

r-vasquez replied:
This was removed and enabled by default :+1:

twmb commented:
This is similar to #14025 but adds the disabled column. I think the two PRs should be merged?

daisukebe commented:
Definitely mine should be merged here. Some comments:

* The new API will only be applicable >= 23.3, I'd suppose. If the command runs against an older version, < 23.3, it will just fail because of the invalid endpoint. Do we want to check the cluster version beforehand like we did in the decommission command earlier?
* A typical use case I wanted to cover in https://github.com/redpanda-data/redpanda/pull/14025 was identifying cores on each node where a given partition replica is assigned on. That is, having the `--partition` flag will be helpful in a case there are hundreds of partitions in a single topic.

r-vasquez commented:
Last[ force push](https://github.com/redpanda-data/redpanda/compare/724662b0c73d3850b8ef219e0488aa63fb79c535..29a2c37cb5e0ff941a696b911a635bcf81556732):

- Addressed initial comments
- Added 'describe' as an alias for the command
- Updated help text
- Removed --with-internals, and made 'show internal topics' the default
- Added --partition flag, to filter results based on the given partitions
- Changed the per-topic requests to be concurrent
- Changed Replica assignment format, from using : to -
- Renamed --disabled to --only-disabled

r-vasquez commented:
I'm adding a new commit later to fall back into the old API when the request fails. (to cover the situation when a new rpk talks to an old cluster)

r-vasquez commented:
Last [Force Push ](https://github.com/redpanda-data/redpanda/compare/29a2c37cb5e0ff941a696b911a635bcf81556732..642c9e2e0fc5fd49680513184b1dfe95fb5ded11):

Introduces a new commit with a new ability to fall back to the previous `/v1/partitions` endpoint in case the `/v1/cluster/partitions` returns a 404. This is useful for users that are running `rpk` to older clusters.

daisukebe commented on src/go/rpk/pkg/cli/cluster/partitions/list.go:
```diff
@@ -1,207 +1,231 @@
+		Use:     "list [TOPICS]",
+		Aliases: []string{"ls", "describe"},
+		Short:   "List partitions in the cluster",
+		Long: `List partitions in the cluster
```
Wonder we should explain what the `DISABLED` column means by along with the meaning of `-`.

r-vasquez replied:
Added a new paragraph for this :+1:

gene-redpanda commented on src/go/rpk/pkg/cli/cluster/partitions/move_status.go:
```diff
@@ -0,0 +1,216 @@
+
+    rpk cluster partitions move-status foo bar --partition 0,1,2
+
+The output contains the following columns, where PARTITION-SIZE is in bytes.
```
```suggestion
The output contains the following columns with PARTITION-SIZE in bytes.
```

r-vasquez replied:
Fixed in a new commit: 8c76b5f62b12ebc5a1affae0971a435c475d18ac

gene-redpanda commented on src/go/rpk/pkg/cli/cluster/partitions/move_status.go:
```diff
@@ -0,0 +1,216 @@
+    BYTES-MOVED
+    BYTES-REMAINING
+
+Using "--print-all / -a" the command additionally prints "RECONCILIATION STATUSES"
+which reveals internal states on how the ongoing reconciliations work for debugging
+purposes. That is, reported errors don't necessarily mean real problems.
```
```suggestion
Using "--print-all / -a" the command additionally prints "RECONCILIATION STATUSES"
which reveals internal states for the ongoing reconciliations for debugging
purposes. Reported errors do not necessarily mean real problems.
```

What does "real problems" mean here, and what is it that we're actually reporting? The last sentence could stand to be a lot clearer in communicating what the takeaway should be for an administrator reviewing the output of this command.

r-vasquez replied:
Hey @daisukebe, my PR modified the name of the `move-status` file that you wrote. Do you have any suggestions on how to rephrase this part?

I'm definitely not an expert on how the internal status of a reconciliation/movement works.

daisukebe replied:
The `RECONCILIATION STATUSES` section prints very detailed / internal state of Raft reconfiguration (aka partition movement). It typically reports something like this but these are common in general. 

```
NODE-ID  CORE  TYPE    RETRY-NUMBER  REVISION  STATUS
1        1     update  20            664       Generic failure occurred during partition operation execution (cluster::errc:52)
2        1     update  20            664       Current node is not a leader for partition (cluster::errc:17)
3        0     update  20            664       Current node is not a leader for partition (cluster::errc:17)
4        1     update  16            664       Current node is not a leader for partition (cluster::errc:17)
```

This will make a user who reads this believe there's a problem at play, but it's not true. Core team needs to improve the wording, but it's not done yet. This was what I meant "do not necessarily mean real problems".

Even though, it's worth reporting the status with `--a` in a case there's something wrong happening and core team wants to review it.

RE: rephrasing,

Hmm, something like this? Wdyt?

current
```
Reported errors do not necessarily mean real problems.
```

new
```
Use the option with advise from Redpanda support.
```

gene-redpanda replied:
"Advice" but otherwise that's great and clearly communicated when to use this. Thanks!

daisukebe replied:
Uses don't need to use it unless Redpanda support asks. How about this?
```
Use the option when Redpanda support requests.
```

gene-redpanda commented on src/go/rpk/pkg/cli/cluster/partitions/move_status.go:
```diff
@@ -0,0 +1,216 @@
+
+	cmd.Flags().BoolVarP(&all, "print-all", "a", false, "Print internal states about movements for debugging")
+	cmd.Flags().BoolVarP(&human, "human-readable", "H", false, "Print the partition size in a human-readable form")
+	cmd.Flags().StringSliceVarP(&partitions, "partition", "p", nil, "Partitions to print the ongoing movements (repeatable)")
```
Can this be clarified? What does "Partitions to print the ongoing movements (repeatable)" mean in this context?

r-vasquez replied:
The command `move-status` prints `all` ongoing movements, the `partition` flag prints the ongoing movements for those partitions.

r-vasquez replied:
Changed this to `Partitions to filter ongoing movements status (repeatable)` in 8c76b5f62b12ebc5a1affae0971a435c475d18ac

gene-redpanda commented on src/go/rpk/pkg/cli/cluster/partitions/list.go:
```diff
@@ -1,207 +1,231 @@
 	}
+	cmd.Flags().BoolVarP(&all, "all", "a", false, "If true, list all partitions in the cluster")
+	cmd.Flags().BoolVar(&disabledOnly, "only-disabled", false, "If true, list disabled partitions only")
+	cmd.Flags().IntSliceVarP(&partitions, "partition", "p", nil, "Partitions to show current assignments (repeatable)")
```
```suggestion
	cmd.Flags().IntSliceVarP(&partitions, "partition", "p", nil, "Partitions to show current assignments (repeatable)")
```

This could stand to be clearer.

r-vasquez replied:
Changed to `List of comma-separated partitions IDs that you wish to filter the results with`

gene-redpanda commented on src/go/rpk/pkg/cli/cluster/partitions/list.go:
```diff
@@ -1,207 +1,231 @@
+				// with an old cluster.
+				if he := (*adminapi.HTTPResponseError)(nil); errors.As(err, &he) {
+					if he.Response.StatusCode == http.StatusNotFound {
+						out.Die("unable to query all partitions in the cluster: %v You may need to upgrade the cluster to access this feature or try listing per-topic with: 'rpk cluster partition list [TOPICS]'", err)
```
This is a great error message

gene-redpanda reviewed (approved):
I had a few comments on the help comments but everything else looked fine to me

r-vasquez commented on src/go/rpk/pkg/cli/cluster/partitions/move_status.go:
```diff
@@ -0,0 +1,216 @@
+
+const helpListMovement = `Show ongoing partition movements.
+
+By default this command lists all the ongoing partition movements in the cluster.
```
Fixed in a new commit: 8c76b5f62b12ebc5a1affae0971a435c475d18ac

r-vasquez commented:
e050fc8a01fcd83f3fd3dad307a9acf1a9388904: 

- Addressed some of the comments, please note that this PR does not introduces the `move-status` command, we renamed that file to move_status.go (previously list.go). Github is showing this as a new file.

I split the commit history to help you with the review and avoid confusions :+1:

twmb commented on src/go/rpk/pkg/cli/cluster/partitions/list.go:
```diff
@@ -43,6 +48,10 @@ internal namespace using the "{namespace}/" prefix.
 replicas assignments in the form of: <Node-ID>-<Core>.
 
+If the DISABLED column contains a '-' value, then it means you are running this
+command against a cluster that is outdated and does not support the underlying 
+API.
```
`against an older cluster that does not support`, or maybe even drop `older`. I don't think outdated has the best connotation here.

twmb commented on src/go/rpk/pkg/cli/cluster/partitions/list.go:
```diff
@@ -80,26 +89,39 @@ List all in json format.
+				// with an old cluster.
+				if he := (*adminapi.HTTPResponseError)(nil); errors.As(err, &he) {
+					if he.Response.StatusCode == http.StatusNotFound {
+						out.Die("unable to query all partitions in the cluster: %v You may need to upgrade the cluster to access this feature or try listing per-topic with: 'rpk cluster partition list [TOPICS]'", err)
```
tiny format nit, change `%v You` to `%v; you`, and maybe add `...` after `TOPICS` (so it looks like `[TOPICS...]` although I guess that would mean a corresponding ... on line R38?

r-vasquez replied:
w.r.t adding the `;` I had this first, however, the admin API error contains a new line at the end so I ended up having:

```
unable to query all partitions in the cluster: request GET http://127.0.0.1:9644/v1/cluster/partitions?with_internal=true&disabled=false failed: Not Found, body: "{\"message\": \"Not found\", \"code\": 404}"
; You may need to upgrade the cluster to access this feature or try listing per-topic with: 'rpk cluster partition list [TOPICS]'

```

That's why I omitted the semicolon. Actually, I think that I need to remove the white space after the `%v`

I'll add the three dots though.

r-vasquez replied:
Fixed, thanks!

twmb commented on src/go/rpk/pkg/cli/cluster/partitions/list.go:
```diff
@@ -80,26 +89,39 @@ List all in json format.
+							if disabledOnly {
+								zap.L().Sugar().Warn("--only-disabled flag is not supported in this API version")
+							}
+							cPartition, err = topicPartitions(egCtx, cl, ns, topicName)
```
Why did #14025 need to also issue a request per-partition? It looks like all the info is available in just the topic req?

r-vasquez replied:
Yes, it's available on the topic req, I guess per-partition is to avoid requesting large topic metadata req, and instead just request the partition info you are interested in.

However, I went with the "issue for the whole topic -> then filter" because I felt like adding this optimization here is more convoluted for only a fallback mechanism that will be fairly unused in a couple of months.

twmb commented on src/go/rpk/pkg/cli/cluster/partitions/list.go:
```diff
@@ -80,26 +89,39 @@ List all in json format.
+							if !isNotFoundErr {
+								return fmt.Errorf("unable to query cluster partition metadata of topic %q: %v", topicName, err)
+							}
+							zap.L().Sugar().Warn("Admin API 'GET /v1/cluster/partitions' returned 404, trying now with 'GET /v1/partitions'")
+							if disabledOnly {
+								zap.L().Sugar().Warn("--only-disabled flag is not supported in this API version")
+							}
```
I'd merge these to one log line, and I would only log this once in the function (it is likely this will fail for all nodes in the cluster, all requests).

Mind changing this to:
* `var logFallbackOnce sync.Once` at top of function
* include the --only-disabled part of the message in _one_ log line (so, have a conditional for which log line to print)

r-vasquez replied:
Done, thanks

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/41196#018bd602-706f-43f9-8c20-a61cbd103666
<!--c3eb6474-d6ea-469c-96ab-a7607e4f2bba-->

daisukebe commented:
@r-vasquez , have you considered about backporting this to 23.2.x?

r-vasquez commented:
@daisukebe Not really, since it was dependent originally on a feature that is coming for v24 and introduces the new "enabling/disabling" partition concept. However, now that we have a fallback we might backport it if needed.

Let me know if it's needed and we can manually backport the pertinent commits.

daisukebe commented:
@r-vasquez yea, that'd be great! The combination of list & move command really makes our operation simpler and easier in the wild.

daisukebe commented:
/backport v23.2.x

vbotbuildovich commented:
Failed to create a backport PR to v23.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-14862-v23.2.x-650 remotes/upstream/v23.2.x
git cherry-pick -x 8c76b5f62b12ebc5a1affae0971a435c475d18ac 2880a07352119dc932a18663d06051f9019b39dd 560b611be6e16255d819aecd925054e8d10bc99c e19a1abc003bfab5517350aa24af0d9b3f158218 96e21632c1d727ecef524ac266be93abce184386 24b5d9b8a7db0d15f288c022bf857b58d7aacdd9 27b165fa73ec0522b2ce186957112551da046215 0d9ce707bbd6b7eb7406894abe1360062525cbd9
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/6945843863).
